test/model/narrow_test.dart

```diff
@@ -22,13 +48,57 @@ void main() {
     });
   });
 
+  group('ChannelNarrow', () {
+    test('containsMessage', () {
+      final stream = eg.stream();
+      final otherStream = eg.stream();
+      final narrow = ChannelNarrow(stream.streamId);
+      check(narrow.containsMessage(
+        eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]))).isFalse();
+      check(narrow.containsMessage(
+        eg.streamMessage(stream: otherStream, topic: 'topic'))).isFalse();
+      check(narrow.containsMessage(
+        eg.streamMessage(stream: stream,      topic: 'topic'))).isTrue();
+
+      check(narrow.containsMessage(
+        _TestDmMessage(allRecipientIds: [1]))).isFalse();
+      check(narrow.containsMessage(
+        _TestStreamMessage(stream: otherStream, topic: 'topic'))).isFalse();
+      check(narrow.containsMessage(
+        _TestStreamMessage(stream: stream,      topic: 'topic'))).isTrue();
+    });
+  });
+
   group('TopicNarrow', () {
     test('ofMessage', () {
       final stream = eg.stream();
       final message = eg.streamMessage(stream: stream);
       final actual = TopicNarrow.ofMessage(message);
       check(actual).equals(TopicNarrow(stream.streamId, message.topic));
     });
+
+    test('containsMessage', () {
+      final stream = eg.stream();
+      final otherStream = eg.stream();
+      final narrow = eg.topicNarrow(stream.streamId, 'topic');
+      check(narrow.containsMessage(
+        eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]))).isFalse();
+      check(narrow.containsMessage(
+        eg.streamMessage(stream: stream,      topic: 'topic2'))).isFalse();
+      check(narrow.containsMessage(
+        eg.streamMessage(stream: otherStream, topic: 'topic'))).isFalse();
+      check(narrow.containsMessage(
+        eg.streamMessage(stream: stream,      topic: 'topic'))).isTrue();
```

Comment:
nit: swap the middle two cases to make the order slightly more logical: a DM is farthest away from belonging in this narrow, then a message to some other channel, then one to a different topic in the same narrow, then finally one to this same conversation